[dynamo] Add BACKEND_MATCH guard to detect and recompile when backend changes

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #107337

**Motivation:**
We try to make torch.cond use torch.compile automatically so that we could error out when there is side-effects in the branches and correctly handle the closures. 

Before this PR, we have a warning if we don't turn on a config raise_on_backend_change (turning it on gives us an error) for the following code:
```python
def foo()


# Inside torch.cond, we'd like to do something like 
torch.compile(foo, backend="eager", fullgraph=True)(...)
...
# Users may then call torch.compile somewhere else. 
# Dynamo will use the cached code of foo for "eager" backend 
# but we expect dynamo to recompile with "inductor" backend.
torch.compile(foo, backend="inductor")(...)
```

This PR adds a BACKEND_MATCH guard. Effectively, it implements a per-backend cache. In the above example, the cached code for "eager" won't work for "inductor" due to guard check failures and the second torch.compile will do a re-compilation. In the future, it might be useful to have something like a configuration guard that guards against dynamo configuration changes across different compiles (e.g. compile a function with fullgraph=False then compile it again with fullgraph=True).

**Implementation:**
1. We add a guarded_backend_cache and check the most_recent_backend against the backend associated with cached code. We also remove the raise_on_backend_change flag.


Note: More lines are printed for debug log due to newly added context manager and guard adds .

**Test Plan:**
Removed original tests that raise on different backend and add a new test to test whether the BACKEND_MATCH guard can guard against backend change.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @anijain2305

ydwu4 committed:
[dynamo] Add BACKEND_MATCH guard to recompile if backend changes

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo] Add BACKEND_MATCH guard to recompile if backend changes"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo] Add BACKEND_MATCH guard to recompile if backend changes"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo] Add BACKEND_MATCH guard to recompile if backend changes"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo] Add BACKEND_MATCH guard to detect and recompile when backend changes"


**Motivation:**
We try to make torch.cond use torch.compile automatically so that we could error out when there is side-effects in the branches and correctly handle the closures. 

Before this PR, we have a warning if we don't turn on a config raise_on_backend_change (turning it on gives us an error) for the following code:
```python
def foo()


# Inside torch.cond, we'd like to do something like 
torch.compile(foo, backend="eager", fullgraph=True)(...)
...
# Users may then call torch.compile somewhere else. 
# Dynamo will use the cached code of foo for "eager" backend 
# but we expect dynamo to recompile with "inductor" backend.
torch.compile(foo, backend="inductor")(...)
```

This PR adds a BACKEND_MATCH guard. Effectively, it implements a per-backend cache. In the above example, the cached code for "eager" won't work for "inductor" due to guard check failures and the second torch.compile will do a re-compilation. In the future, it might be useful to have something like a configuration guard that guards against dynamo configuration changes across different compiles (e.g. compile a function with fullgraph=False then compile it again with fullgraph=True).

**Implementation:**
We add a guarded_backend_cache and check the most_recent_backend against the backend associated with cached code. We also remove the raise_on_backend_change flag.

**Test Plan:**
Removed original tests that raise on different backend and add a new test to test whether the BACKEND_MATCH guard can guard against backend change.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

ezyang commented on torch/_dynamo/eval_frame.py:
```diff
@@ -410,15 +408,8 @@ def __init__(
-                        "calling `torch._dynamo.reset()` to take effect"
-                    )
             most_recent_backend = compiler_fn
+            guarded_backend_cache[id(most_recent_backend)] = most_recent_backend
```
It seems like it would be better to have a thread local which is set when we enter a torch.compile region specifying what the current backend is. At the very least it needs to be thread local, which it doesn't seem like `most_recent_backend` is. (Yes, I know some of this is preexisting problems, but now you know.)

ydwu4 replied:
I've modified accordingly and added a test for multi-threads but I'm not sure if the implementation is good enough.

ydwu4 committed:
Update on "[dynamo] Add BACKEND_MATCH guard to detect and recompile when backend changes"


**Motivation:**
We try to make torch.cond use torch.compile automatically so that we could error out when there is side-effects in the branches and correctly handle the closures. 

Before this PR, we have a warning if we don't turn on a config raise_on_backend_change (turning it on gives us an error) for the following code:
```python
def foo()


# Inside torch.cond, we'd like to do something like 
torch.compile(foo, backend="eager", fullgraph=True)(...)
...
# Users may then call torch.compile somewhere else. 
# Dynamo will use the cached code of foo for "eager" backend 
# but we expect dynamo to recompile with "inductor" backend.
torch.compile(foo, backend="inductor")(...)
```

This PR adds a BACKEND_MATCH guard. Effectively, it implements a per-backend cache. In the above example, the cached code for "eager" won't work for "inductor" due to guard check failures and the second torch.compile will do a re-compilation. In the future, it might be useful to have something like a configuration guard that guards against dynamo configuration changes across different compiles (e.g. compile a function with fullgraph=False then compile it again with fullgraph=True).

**Implementation:**
We add a guarded_backend_cache and check the most_recent_backend against the backend associated with cached code. We also remove the raise_on_backend_change flag.

**Test Plan:**
Removed original tests that raise on different backend and add a new test to test whether the BACKEND_MATCH guard can guard against backend change.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo] Add BACKEND_MATCH guard to detect and recompile when backend changes"


**Motivation:**
We try to make torch.cond use torch.compile automatically so that we could error out when there is side-effects in the branches and correctly handle the closures. 

Before this PR, we have a warning if we don't turn on a config raise_on_backend_change (turning it on gives us an error) for the following code:
```python
def foo()


# Inside torch.cond, we'd like to do something like 
torch.compile(foo, backend="eager", fullgraph=True)(...)
...
# Users may then call torch.compile somewhere else. 
# Dynamo will use the cached code of foo for "eager" backend 
# but we expect dynamo to recompile with "inductor" backend.
torch.compile(foo, backend="inductor")(...)
```

This PR adds a BACKEND_MATCH guard. Effectively, it implements a per-backend cache. In the above example, the cached code for "eager" won't work for "inductor" due to guard check failures and the second torch.compile will do a re-compilation. In the future, it might be useful to have something like a configuration guard that guards against dynamo configuration changes across different compiles (e.g. compile a function with fullgraph=False then compile it again with fullgraph=True).

**Implementation:**
We add a guarded_backend_cache and check the most_recent_backend against the backend associated with cached code. We also remove the raise_on_backend_change flag.

**Test Plan:**
Removed original tests that raise on different backend and add a new test to test whether the BACKEND_MATCH guard can guard against backend change.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

ydwu4 commented on torch/_dynamo/eval_frame.py:
```diff
@@ -651,14 +651,13 @@ def guard_export_print(guards):
-            )(f)
-            # TODO(voz): We may have instances of `f` that mutate inputs, we should track sideffects and reject.
-            opt_f(*args, **kwargs)
+        opt_f = optimize(
```
I assume this most_recent_backend patch is to disable the _different_backend check previously. It seems not useful after this PR so remove it. Same for export.

ydwu4 commented on torch/_dynamo/eval_frame.py:
```diff
@@ -1176,7 +1175,7 @@ def result_capturing_wrapper(*graph_inputs):
         if tracing_mode != "symbolic":
             assume_static_by_default = True
-        with patch(f"{__name__}.most_recent_backend", None), config.patch(
+        with config.patch(
```
Same reasoning as above.

ydwu4 committed:
Update on "[dynamo] Add BACKEND_MATCH guard to detect and recompile when backend changes"


**Motivation:**
We try to make torch.cond use torch.compile automatically so that we could error out when there is side-effects in the branches and correctly handle the closures. 

Before this PR, we have a warning if we don't turn on a config raise_on_backend_change (turning it on gives us an error) for the following code:
```python
def foo()


# Inside torch.cond, we'd like to do something like 
torch.compile(foo, backend="eager", fullgraph=True)(...)
...
# Users may then call torch.compile somewhere else. 
# Dynamo will use the cached code of foo for "eager" backend 
# but we expect dynamo to recompile with "inductor" backend.
torch.compile(foo, backend="inductor")(...)
```

This PR adds a BACKEND_MATCH guard. Effectively, it implements a per-backend cache. In the above example, the cached code for "eager" won't work for "inductor" due to guard check failures and the second torch.compile will do a re-compilation. In the future, it might be useful to have something like a configuration guard that guards against dynamo configuration changes across different compiles (e.g. compile a function with fullgraph=False then compile it again with fullgraph=True).

**Implementation:**
We add a guarded_backend_cache and check the most_recent_backend against the backend associated with cached code. We also remove the raise_on_backend_change flag.

**Test Plan:**
Removed original tests that raise on different backend and add a new test to test whether the BACKEND_MATCH guard can guard against backend change.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

ezyang commented on torch/_dynamo/__init__.py:
```diff
@@ -62,8 +62,12 @@ def reset() -> None:
-    if hasattr(eval_frame.most_recent_backend, "reset"):
-        eval_frame.most_recent_backend.reset()
-    eval_frame.most_recent_backend = None
+    cached_backends = getattr(eval_frame.guarded_backend_cache, "cached_backends", None)
```
What's... this?

ydwu4 replied:
guarded_backend_cache holds a dictionary of id2backends and a most recent backend. People sometimes directly call torch._dynamo.reset() before compiling any function so we have to check whether eval_frame.guarded_backend_cahe have cached_backends or not. Also because guarded_backend_cache is a thread local, we cannot initializes it globally for all threads beforehand so we have to rely on this getattr checks

ezyang commented on torch/_dynamo/guards.py:
```diff
@@ -75,6 +75,16 @@
+        ),
+        (
+            "___lookup_backend",
+            lambda backend_obj_id: torch._dynamo.eval_frame.guarded_backend_cache.cached_backends[
```
Can you explain why all of this code exists? It is not obvious to me. Naively, I expect a backend object to be a plain singleton object which I do not mind if it is kept live by cache entries, and I can just hold onto it directly and do an `is` test on it.

ydwu4 replied:
Ok! We want to use "==" to compare the backend instead of "is" because 1. every time we call torch.compile(backend="inductor"), we actually creates a new _TorchCompileInductorWrapper object (here: https://github.com/pytorch/pytorch/blob/main/torch/__init__.py#L1715). 2. The previous _different_backend check uses "==" instead of "is", which I think it's reasonable since "is" cannot detect obj mutations. 


When we **check** the BACKEND_MATCH guard, we want to compare current backend with the backend associated with current cache. That backend should be baked in the check function of the cache as a constant when we **compiling** the BACKEND_MATCH guard. This pr baked in the id(backend) into the check function. We then need the cached_backends dictionary to map the id back to the original backend object so that we can compare with "==".

voznesenskym replied:
(1) _TorchCompileInductorWrapper already has a defined `__eq__` it does not seem a stretch to just have it define `__hash__` and not hack around it like this, and move over to using `is`?

ydwu4 replied:
Thanks for the comment! I considered the alternative of implementing the guard check function based on '__hash__', but  '__hash__' doesn't prevent collisions in its spirit: even If the hash of the backend happens to be the same we're not certain that these two backend will do the same thing.

'is' doesn't check '__eq__' or '__hash__', it just checks whether these two are the same object in memory. Using 'is' will have the lambda and ephemeral wrapper problem that @jansel mentions.

jansel commented:
One concern I have about this is code like:
```
for _ in range(100):
  y=torch.compile(foo, backend=lambda gm, _: gm)(x)
```
In this case the `id()` of the backend could be different each time since the lambda is a temporary object.  

Similarly, for many backends we create wrapper object instances to encode options.  The wrapper objects can often be ephemeral.

I haven't taken a close look at the code yet, but can you explain how these are addressed?

ydwu4 commented:
This PR uses "==" to compare the current backend with the backend associated with cached code. When creating the guard, we bakes in id(backend) only for retrieval of the backend and the real guard check is done by "__current_backend == look_up(id)". This will work if the backend properly implements "\_\_eq\_\_".

In the case of lambdas, "==" falls back to id(), which fails the guard in different iterations. We don't have a good solution for it.

In the case of wrapper objects, if "\_\_eq\_\_" is implemented properly (as in _TorchCompileInductorWrapper), we don't need to recompile every time.

voznesenskym commented on torch/_dynamo/guards.py:
```diff
@@ -514,6 +524,14 @@ def DEFAULT_DEVICE(self, guard: Guard):
             guard, [f"utils_device.CURRENT_DEVICE == {m.CURRENT_DEVICE!r}"]
         )
 
+    def BACKEND_MATCH(self, guard: Guard):
```
We don't want more than 1 of these. We also need an invariant to ensure we don't somehow cross wires and try to accumulate guards across different backends for the same guard fn.

I would change this to be a flag that says "Yea, we need a backend match w/ backend FOO" and when we go to produce and install the check_fn, that's where we add this to the very front of the guard set. If another flag is raised, where the backend does not match FOO in our example, we need to raise.

voznesenskym replied:
We also need to be very careful that identity is good enough here - there's a reason we were toying with the idea of "`'eager'` is special" - because it avoids a whole class of cache destroying problems where a seemingly mismatched identity, for a functionally identical backend, means we never cache hit.

ydwu4 replied:
> We also need an invariant to ensure we don't somehow cross wires and try to accumulate guards across different backends for the same guard fn.

It seems to me the invariant is ready: BACKEND_GUARD is added to the tracing_context in OutputGraph's constructor. So it is 1-1 mapping with the root instruction translator. So for a frame, the first time dynamo encounter this frame, it'll write this guard into the check_fn when dynamo successfully compiles a graph. Later on, if we encountered it again but with a different backend, we'll do a recompilation.

> We also need to be very careful that identity is good enough

Yes, I agree. Jason's comment mentions two cases. And we don't have a good way for lambda problem right now and the `== ` can address the wrapper problem for the wrappers that we have now.

voznesenskym replied:
This will add BACKEND_GUARD as many times as it encounters it. As I said above, there should only be a single one. And that single one should be at the very fore of the check_fn. Please replace this code to not append guard code, but to set a flag instead, which then instructs us how to write the code to the front of check_fn.

ydwu4 replied:
Ok, let me try to digest a bit. So you're suggesting that we should keep an 1-1 mapping between backend and original code object and raise exceptions if the code object is compiled with a different backend. Specifically, we keep the backend that dynamo uses to compile this code object for the first time. If later on, any torch.compile is applied on this code object again but with a different backend, we should raise an exception. Is this the proposal?

I have some concerns about this single-cache design: 
1. @jansel 's the lambda and ephemeral wrapper case would not be possible: in these cases, we expect different lambdas or wrappers to share the optimized code object instead of raising on exception. Or, if we somehow can make sharing possible (e.g. for wrapper, we checkwith `==` instead of `is`), I would argue that we could do the same for current design to avoid repeated compilation.
2. We still need to figure out how to implement a separate "eager" cache for torch.cond to support automatic dynamo inspection. 
3. How do we scale this design to compile configurations beyond backend? e.g. fullgraph, capture_dynamic_output_shape_ops etc. Even for the same "eager" backend, these configuration changes may need a separate cache to avoid incorrect behaviors.

voznesenskym replied:
> Ok, let me try to digest a bit. So you're suggesting that we should keep an 1-1 mapping between backend and original code object and raise exceptions if the code object is compiled with a different backend. Specifically, we keep the backend that dynamo uses to compile this code object for the first time. If later on, any torch.compile is applied on this code object again but with a different backend, we should raise an exception. Is this the proposal?

No, this is not the proposal, at all. Let me start again.

With the code as you have written it, your `BACKEND_MATCH` will add code over and over to the guards, for every BACKEND_MATCH we accumulate during trace. This is wasteful. Imagine stumbling upon a `cond` during trace incurs a `BACKEND_MATCH` - and let's say we have 4 conds in the traced program. With how this PR is written, we will have 4 identical BACKEND_MATCH. This is wasteful, and also fails to protect a useful invariant (that, which, for a given traced frame, the backend in BACKEND_MATCH is always the same).

**Current code:**

A BACKEND_MATCH that adds backend matching code for every BACKEND_MATCH in the frame we are tracing

**My proposal:**

A BACKEND_MATCH that sets a flag upon discovery - subsequent accumulations of BACKEND_MATCH just check that that the backend is the same as found in earlier BACKEND_MATCH's accumulated during this frame. If this flag is set, we add a single backend comparison to the front of the check fn

**Proposed test:**

Write code that accumulates K BACKEND_MATCHs, where K > 1, and then other guards as well

First compile: BACKEND_MATCH + other guards
Second compile: Recompile due to backend change AND tensor change

Assert that (1) there is only 1 backend comparison in the guard fn (despite having K accumulations) and (2) that the failure was because of backend mismatch (proving this guard was moved to the very front of the check fn)

ydwu4 replied:
> Imagine stumbling upon a cond during trace incurs a BACKEND_MATCH - and let's say we have 4 conds in the traced program. With how this PR is written, we will have 4 identical BACKEND_MATCH.

No, we'll only create 1 BACKEND_MATCH guard and match it 3 times. See below example based on current PR and the dynamo logs:
```python
import torch
from torch._dynamo.testing import CompileCounter

cnt = CompileCounter()
x = torch.randn([3, 4])
def cond(x):
    def k(x):
        return x.sin(), x.cos()
    return torch.compile(k, backend=cnt)(x)

def g(x):
    return cond(x), cond(x), cond(x), cond(x)

g(x)
print(f"frame_count:{cnt.frame_count} op_count:{cnt.op_count}")
```
logs: https://www.internalfb.com/phabricator/paste/view/P811076827

voznesenskym replied:
> This can happen when we have graph breaks

This is a new frame, so will get a new check_fn.

> So is this the case that you refer to? 

No

> No, we'll only create 1 BACKEND_MATCH guard and match it 3 times

This is incidentally the case because you invoke it once, but the function is written in a way that does not protect this invariant. We can write it better. I explained how above.

ydwu4 replied:
In the case of cond, that's the exact pattern, right? I don't see it as an incident. Can you show me an example of multiple BACKEND_MATCH guards in a program? I haven't been able to come up with one...

voznesenskym replied:
My point is, anyone can go and add it. This PR doesnt make that mistake, but as the function is written, it can happen. Just do `add_guard(GuardBuilder.BACKEND_MATCH)`, and bam, there's your bug. A well written version of this guard makes that impossible, and is not really any more complicated than what you have today, AND has the added bonus of moving this guard to the forefront of the check_fn.

ydwu4 replied:
Yeah, I can buy the argument that someone might do a random `add_guard(GuardBuilder.BACKEND_MATCH)` somewhere. I'll wait for more comment and modify it accordingly. Thanks!

ezyang replied:
Voz, I don't understand the point you are making here. Guards are tossed in a set so they will get deduped anyway. And we don't actually have any precedent for what you suggested; all of the "do it once" guards in OutputGraph are done in the same way this guard was done

voznesenskym replied:
No - those are global state guards - for things we cannot discover lazily / AKA need ambient global state protection. DETERMINISTIC_ALGORITHMS, GRAD_MODE, DEFAULT_DEVICE, TORCH_FUNCTION_STATE

This is a little different. 

1) There is no reason to accrue this guard if we did not hit upon a higher order op that always invoke PT2, even not under a compile region. My understanding was that this was entirely for `cond` operators.

If the goal is to have a per-backend cache, and we want this available generally, then this isn't the way to do it IMO.

2) This guard should be at the front of the other guards, I dont think we have a good mechanism for this other than the one I proposed.

IDC too much though, I can open a PR with my proposed changes if we decide to land this. It's not blocking, just less than ideal.

ydwu4 replied:
For 1, I want add that this pr is more than `entirely for cond`. In fact, I've triggered an silent failure while developing this PR. 

For example, we have [tests] that rely on first compile a function with one backend then compile a test with another backend ("aot_eager" and "inductor").https://github.com/pytorch/pytorch/blob/e0f1fe102ac85bcf3b7b01e50fa56df1bcbbfef1/torch/testing/_internal/optests/compile_check.py#L47-L48

Before this PR, we never thought about the second torch.compile with "inductor" backend doesn't trigger compilation due to lacking of a BACKEND_MATCH guard.

voznesenskym reviewed (changes requested):
See comments

ezyang commented:
Based on jansel's comments, we will either have to give up on this, or do something more complex.

The most logical way to deal with the problem of lambda is to distinguish between backends which support equality and those that don't. We would need to do legacy behavior for the latter

voznesenskym commented:
> Based on jansel's comments, we will either have to give up on this, or do something more complex.
> 
> The most logical way to deal with the problem of lambda is to distinguish between backends which support equality and those that don't. We would need to do legacy behavior for the latter

This is part of why I wanted the "eager is special" approach. Alternatively, custom backend fns/lambdas/etc are out - and we only backend match on known registered backends.

ydwu4 commented:
I can implement Ed's approach: distinguish between backends which support equality and those that don't. Do legacy behavior for the latter. Should I start doing it? @jansel

jansel commented:
Yeah, I think some sort of legacy case would work.  For all the builtin backends we should be able to do equality comparison.

jansel commented:
On further though, maybe we should just eat the BC-breaking change and start saying that multiple calls to torch.compile() could lead to recompiles in some cases they didn't in v2.0.

jansel commented on torch/_dynamo/eval_frame.py:
```diff
@@ -409,16 +418,7 @@ def __init__(
-                        "calling `torch._dynamo.reset()` to take effect"
-                    )
-            most_recent_backend = compiler_fn
+            _set_current_backend(compiler_fn)
```
I don't think this will work correctly with nesting.  We may need to store the prior backend and restore it on exit.

ydwu4 committed:
Update on "[dynamo] Add BACKEND_MATCH guard to detect and recompile when backend changes"


**Motivation:**
We try to make torch.cond use torch.compile automatically so that we could error out when there is side-effects in the branches and correctly handle the closures. 

Before this PR, we have a warning if we don't turn on a config raise_on_backend_change (turning it on gives us an error) for the following code:
```python
def foo()


# Inside torch.cond, we'd like to do something like 
torch.compile(foo, backend="eager", fullgraph=True)(...)
...
# Users may then call torch.compile somewhere else. 
# Dynamo will use the cached code of foo for "eager" backend 
# but we expect dynamo to recompile with "inductor" backend.
torch.compile(foo, backend="inductor")(...)
```

This PR adds a BACKEND_MATCH guard. Effectively, it implements a per-backend cache. In the above example, the cached code for "eager" won't work for "inductor" due to guard check failures and the second torch.compile will do a re-compilation. In the future, it might be useful to have something like a configuration guard that guards against dynamo configuration changes across different compiles (e.g. compile a function with fullgraph=False then compile it again with fullgraph=True).

**Implementation:**
We add a guarded_backend_cache and check the most_recent_backend against the backend associated with cached code. We also remove the raise_on_backend_change flag.

**Test Plan:**
Removed original tests that raise on different backend and add a new test to test whether the BACKEND_MATCH guard can guard against backend change.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo] Add BACKEND_MATCH guard to detect and recompile when backend changes"


**Motivation:**
We try to make torch.cond use torch.compile automatically so that we could error out when there is side-effects in the branches and correctly handle the closures. 

Before this PR, we have a warning if we don't turn on a config raise_on_backend_change (turning it on gives us an error) for the following code:
```python
def foo()


# Inside torch.cond, we'd like to do something like 
torch.compile(foo, backend="eager", fullgraph=True)(...)
...
# Users may then call torch.compile somewhere else. 
# Dynamo will use the cached code of foo for "eager" backend 
# but we expect dynamo to recompile with "inductor" backend.
torch.compile(foo, backend="inductor")(...)
```

This PR adds a BACKEND_MATCH guard. Effectively, it implements a per-backend cache. In the above example, the cached code for "eager" won't work for "inductor" due to guard check failures and the second torch.compile will do a re-compilation. In the future, it might be useful to have something like a configuration guard that guards against dynamo configuration changes across different compiles (e.g. compile a function with fullgraph=False then compile it again with fullgraph=True).

**Implementation:**
We add a guarded_backend_cache and check the most_recent_backend against the backend associated with cached code. We also remove the raise_on_backend_change flag.

**Test Plan:**
Removed original tests that raise on different backend and add a new test to test whether the BACKEND_MATCH guard can guard against backend change.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo] Add BACKEND_MATCH guard to detect and recompile when backend changes"


**Motivation:**
We try to make torch.cond use torch.compile automatically so that we could error out when there is side-effects in the branches and correctly handle the closures. 

Before this PR, we have a warning if we don't turn on a config raise_on_backend_change (turning it on gives us an error) for the following code:
```python
def foo()


# Inside torch.cond, we'd like to do something like 
torch.compile(foo, backend="eager", fullgraph=True)(...)
...
# Users may then call torch.compile somewhere else. 
# Dynamo will use the cached code of foo for "eager" backend 
# but we expect dynamo to recompile with "inductor" backend.
torch.compile(foo, backend="inductor")(...)
```

This PR adds a BACKEND_MATCH guard. Effectively, it implements a per-backend cache. In the above example, the cached code for "eager" won't work for "inductor" due to guard check failures and the second torch.compile will do a re-compilation. In the future, it might be useful to have something like a configuration guard that guards against dynamo configuration changes across different compiles (e.g. compile a function with fullgraph=False then compile it again with fullgraph=True).

**Implementation:**
We add a guarded_backend_cache and check the most_recent_backend against the backend associated with cached code. We also remove the raise_on_backend_change flag.

**Test Plan:**
Removed original tests that raise on different backend and add a new test to test whether the BACKEND_MATCH guard can guard against backend change.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo] Add BACKEND_MATCH guard to detect and recompile when backend changes"


**Motivation:**
We try to make torch.cond use torch.compile automatically so that we could error out when there is side-effects in the branches and correctly handle the closures. 

Before this PR, we have a warning if we don't turn on a config raise_on_backend_change (turning it on gives us an error) for the following code:
```python
def foo()


# Inside torch.cond, we'd like to do something like 
torch.compile(foo, backend="eager", fullgraph=True)(...)
...
# Users may then call torch.compile somewhere else. 
# Dynamo will use the cached code of foo for "eager" backend 
# but we expect dynamo to recompile with "inductor" backend.
torch.compile(foo, backend="inductor")(...)
```

This PR adds a BACKEND_MATCH guard. Effectively, it implements a per-backend cache. In the above example, the cached code for "eager" won't work for "inductor" due to guard check failures and the second torch.compile will do a re-compilation. In the future, it might be useful to have something like a configuration guard that guards against dynamo configuration changes across different compiles (e.g. compile a function with fullgraph=False then compile it again with fullgraph=True).

**Implementation:**
1. We add a guarded_backend_cache and check the most_recent_backend against the backend associated with cached code. We also remove the raise_on_backend_change flag.

2. Then newly added context manager and guard adds more lines for debug log.

**Test Plan:**
Removed original tests that raise on different backend and add a new test to test whether the BACKEND_MATCH guard can guard against backend change.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge

ydwu4 commented:
@pytorchbot merge

ydwu4 commented:
@pytorchbot merge -i

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change but inductor perf smoke test starts to regress after this' -c ignoresignal

Here is an example failure https://hud.pytorch.org/pytorch/pytorch/commit/1a64ec7dd48408d6839a5c2cceb55b0c4be2243b

```
2023-09-07T23:51:08.8980334Z + python benchmarks/dynamo/check_hf_bert_perf_csv.py -f /var/lib/jenkins/workspace/test/test-reports/inductor_training_smoketest.csv
2023-09-07T23:51:09.3421344Z hf_Bert                            0.997172
2023-09-07T23:51:09.3421670Z 
2023-09-07T23:51:09.3421819Z Error 1 models performance regressed
2023-09-07T23:51:09.3422217Z     hf_Bert
```

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change but inductor perf smoke test starts to regress after this' -c ignoredsignal

ydwu4 committed:
Update on "[dynamo] Add BACKEND_MATCH guard to detect and recompile when backend changes"


**Motivation:**
We try to make torch.cond use torch.compile automatically so that we could error out when there is side-effects in the branches and correctly handle the closures. 

Before this PR, we have a warning if we don't turn on a config raise_on_backend_change (turning it on gives us an error) for the following code:
```python
def foo()


# Inside torch.cond, we'd like to do something like 
torch.compile(foo, backend="eager", fullgraph=True)(...)
...
# Users may then call torch.compile somewhere else. 
# Dynamo will use the cached code of foo for "eager" backend 
# but we expect dynamo to recompile with "inductor" backend.
torch.compile(foo, backend="inductor")(...)
```

This PR adds a BACKEND_MATCH guard. Effectively, it implements a per-backend cache. In the above example, the cached code for "eager" won't work for "inductor" due to guard check failures and the second torch.compile will do a re-compilation. In the future, it might be useful to have something like a configuration guard that guards against dynamo configuration changes across different compiles (e.g. compile a function with fullgraph=False then compile it again with fullgraph=True).

**Implementation:**
1. We add a guarded_backend_cache and check the most_recent_backend against the backend associated with cached code. We also remove the raise_on_backend_change flag.

2. Then newly added context manager and guard adds more lines for debug log so we change the uppper limit from 50 to 55.

**Test Plan:**
Removed original tests that raise on different backend and add a new test to test whether the BACKEND_MATCH guard can guard against backend change.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo] Add BACKEND_MATCH guard to detect and recompile when backend changes"


**Motivation:**
We try to make torch.cond use torch.compile automatically so that we could error out when there is side-effects in the branches and correctly handle the closures. 

Before this PR, we have a warning if we don't turn on a config raise_on_backend_change (turning it on gives us an error) for the following code:
```python
def foo()


# Inside torch.cond, we'd like to do something like 
torch.compile(foo, backend="eager", fullgraph=True)(...)
...
# Users may then call torch.compile somewhere else. 
# Dynamo will use the cached code of foo for "eager" backend 
# but we expect dynamo to recompile with "inductor" backend.
torch.compile(foo, backend="inductor")(...)
```

This PR adds a BACKEND_MATCH guard. Effectively, it implements a per-backend cache. In the above example, the cached code for "eager" won't work for "inductor" due to guard check failures and the second torch.compile will do a re-compilation. In the future, it might be useful to have something like a configuration guard that guards against dynamo configuration changes across different compiles (e.g. compile a function with fullgraph=False then compile it again with fullgraph=True).

**Implementation:**
1. We add a guarded_backend_cache and check the most_recent_backend against the backend associated with cached code. We also remove the raise_on_backend_change flag.

2. Then newly added context manager and guard adds more lines for debug log so we change the uppper limit from 50 to 55.

**Test Plan:**
Removed original tests that raise on different backend and add a new test to test whether the BACKEND_MATCH guard can guard against backend change.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo] Add BACKEND_MATCH guard to detect and recompile when backend changes"


**Motivation:**
We try to make torch.cond use torch.compile automatically so that we could error out when there is side-effects in the branches and correctly handle the closures. 

Before this PR, we have a warning if we don't turn on a config raise_on_backend_change (turning it on gives us an error) for the following code:
```python
def foo()


# Inside torch.cond, we'd like to do something like 
torch.compile(foo, backend="eager", fullgraph=True)(...)
...
# Users may then call torch.compile somewhere else. 
# Dynamo will use the cached code of foo for "eager" backend 
# but we expect dynamo to recompile with "inductor" backend.
torch.compile(foo, backend="inductor")(...)
```

This PR adds a BACKEND_MATCH guard. Effectively, it implements a per-backend cache. In the above example, the cached code for "eager" won't work for "inductor" due to guard check failures and the second torch.compile will do a re-compilation. In the future, it might be useful to have something like a configuration guard that guards against dynamo configuration changes across different compiles (e.g. compile a function with fullgraph=False then compile it again with fullgraph=True).

**Implementation:**
1. We add a guarded_backend_cache and check the most_recent_backend against the backend associated with cached code. We also remove the raise_on_backend_change flag.


Note: More lines are printed for debug log due to newly added context manager and guard adds .

**Test Plan:**
Removed original tests that raise on different backend and add a new test to test whether the BACKEND_MATCH guard can guard against backend change.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo] Add BACKEND_MATCH guard to detect and recompile when backend changes"


**Motivation:**
We try to make torch.cond use torch.compile automatically so that we could error out when there is side-effects in the branches and correctly handle the closures. 

Before this PR, we have a warning if we don't turn on a config raise_on_backend_change (turning it on gives us an error) for the following code:
```python
def foo()


# Inside torch.cond, we'd like to do something like 
torch.compile(foo, backend="eager", fullgraph=True)(...)
...
# Users may then call torch.compile somewhere else. 
# Dynamo will use the cached code of foo for "eager" backend 
# but we expect dynamo to recompile with "inductor" backend.
torch.compile(foo, backend="inductor")(...)
```

This PR adds a BACKEND_MATCH guard. Effectively, it implements a per-backend cache. In the above example, the cached code for "eager" won't work for "inductor" due to guard check failures and the second torch.compile will do a re-compilation. In the future, it might be useful to have something like a configuration guard that guards against dynamo configuration changes across different compiles (e.g. compile a function with fullgraph=False then compile it again with fullgraph=True).

**Implementation:**
1. We add a guarded_backend_cache and check the most_recent_backend against the backend associated with cached code. We also remove the raise_on_backend_change flag.


Note: More lines are printed for debug log due to newly added context manager and guard adds .

**Test Plan:**
Removed original tests that raise on different backend and add a new test to test whether the BACKEND_MATCH guard can guard against backend change.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge
